pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +55,83 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+type HeadersLazyLoaded struct {
+	Path   string
+	UserID string
+}
+
+type HeadersLazyLoadedTrackerState struct {
+	LazyLoadedBlocks map[string]int64 `json:"lazy_loaded_blocks"`
+	UserID           string           `json:"user_id"`
+}
+
+func (p *ReaderPool) persist(state HeadersLazyLoadedTrackerState, finalPath string) error {
+	// Create temporary path for fsync
+	tmpPath, err := os.CreateTemp("", "lazy-loaded")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpPath.Name())
+
+	data, err := json.MarshalIndent(state, "", "")
+	if err != nil {
+		return err
+	}
+
+	merr := multierror.New()
+	_, err = tmpPath.Write(data)
+	merr.Add(err)
+	merr.Add(tmpPath.Sync())
+	merr.Add(tmpPath.Close())
+
+	if err := merr.Err(); err != nil {
+		return err
+	}
+
+	// fsync the directory too
```

Comment:
can you explain why this is necessary?

Reply:
I am following approach in [shutdown marker code](https://github.com/grafana/mimir/blob/06ed798ad991e1939afd350ad7abbd9e9fedf2a1/pkg/util/shutdownmarker/shutdown_marker.go#L18). I will add the comment (or extracting the logic to a util function) to explain the reason.

Reply:
I think Charles' [suggestion](https://github.com/grafana/mimir/pull/5095#discussion_r1239252646) to extract the code in a shared function makes more sense.

Reply:
I will extract the code in the different PR after this merged.

Reply:
Why not extract it as part of this PR?